src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php

```diff
@@ -3,115 +3,161 @@
+    }
 
     public function persist(Message ...$messages): void
     {
         if (count($messages) === 0) {
             return;
         }
 
-        $sql = "INSERT INTO {$this->tableName} (event_id, aggregate_root_id, version, payload) VALUES ";
-        $values = [];
-        $parameters = [];
+        $insertColumns = [
+            $this->tableSchema->eventIdColumn(),
+            $this->tableSchema->aggregateRootIdColumn(),
+            $this->tableSchema->versionColumn(),
+            $this->tableSchema->payloadColumn(),
+            ...array_keys($this->tableSchema->additionalColumns()),
+        ];
+
+        $insertValues = [];
+        $insertParameters = [];
 
         foreach ($messages as $index => $message) {
             $payload = $this->serializer->serializeMessage($message);
-            $eventIdColumn = 'event_id_' . $index;
-            $aggregateRootIdColumn = 'aggregate_root_id_' . $index;
-            $versionColumn = 'version_' . $index;
-            $payloadColumn = 'payload_' . $index;
-            $values[] = "(:{$eventIdColumn}, :{$aggregateRootIdColumn}, :{$versionColumn}, :{$payloadColumn})";
-            $parameters[$versionColumn] = $payload['headers'][Header::AGGREGATE_ROOT_VERSION] ?? 0;
-            $payload['headers'][Header::EVENT_ID] = $payload['headers'][Header::EVENT_ID] ?? Uuid::uuid4()->toString();
-            $parameters[$eventIdColumn] = $this->uuidToBinary($payload['headers'][Header::EVENT_ID]);
-            $parameters[$payloadColumn] = json_encode($payload, $this->jsonEncodeOptions);
-            $parameters[$aggregateRootIdColumn] = $this->uuidToBinary($payload['headers'][Header::AGGREGATE_ROOT_ID] ?? '');
+            $payload['headers'][Header::EVENT_ID] ??= Uuid::uuid4()->toString();
+
+            $messageParameters = [
+                $this->indexParameter(name: 'event_id', index: $index) => $this->uuidEncoder->encodeString($payload['headers'][Header::EVENT_ID]),
+                $this->indexParameter(name: 'aggregate_root_id', index: $index) => $this->uuidEncoder->encodeString($payload['headers'][Header::AGGREGATE_ROOT_ID]),
+                $this->indexParameter(name: 'version', index: $index) => $payload['headers'][Header::AGGREGATE_ROOT_VERSION] ?? 0,
+                $this->indexParameter(name: 'payload', index: $index) => json_encode($payload, $this->jsonEncodeOptions),
+            ];
+
+            foreach ($this->tableSchema->additionalColumns() as $column => $header) {
+                $messageParameters[$this->indexParameter(name: $column, index: $index)] = $payload['headers'][$header];
+            }
+
+            // Creates a values line like: (:event_id_1, :aggregate_root_id_1, ...)
+            $insertValues[] = implode(', ', $this->nameParameters(array_keys($messageParameters)));
+
+            // Flatten the message parameters into the query parameters
+            $insertParameters = array_merge($insertParameters, $messageParameters);
```

Comment:
I wish there was a better way to do this...

Reply:
Shouldn't this be an `array_push($insertParameters, ...$messageParameters)`? We want to append, not substitute, rigth?

Reply:
No, because the parameters are named.

Reply:
Ah right, I scanned too quickly.